server/service/async/async.go

```diff
@@ -105,7 +106,7 @@ func storePurgeActiveHostID(pool fleet.RedisPool, zsetKey string, hid uint, repo
     return redis.call('ZREMRANGEBYSCORE', KEYS[1], '-inf', ARGV[3])
   `)
 
-	conn := pool.Get()
+	conn := pool.Get().(apmredigo.Conn).WithContext(ctx)
```

Comment:
> conn := pool.Get().(apmredigo.Conn).WithContext(ctx)
 
This type conversion will not work in cluster mode, it won't return an `ampredigo.Conn` at this point (nor when using a `RetryConn` as returned by `ConfigureDoer` when following cluster redirections is enabled). I'm not totally sure how it'd work, haven't looked enough into it right now, but assuming we'll want that apmredigo everywhere we use a redis conn, we should do this wrapping inside the `redis` package - i.e. maybe change Get to be `pool.Get(ctx)` and do the proper wrapping inside that call, instead of every place we need a redis conn.

Reply:
(And we could pass the tracing enabled flag when creating the redis pool instance, so we wouldn't need a global reference).

Reply:
Yeah, I know it won't work which is why its an RFC :). As for using a `pool.Get(ctx)` pattern, I think that will be a requirement as go becomes more context based. Most libraries are changing to take in a context for this reason.

The main concern I have is two fold:

1. The redis library we use (written by you) is not just used by us. So making significant changes to remain 'globally pure' might have much farther reaching consequences.
2. While I agree with global purity, the very next change would be to enable logging in this area of code, which unless I am not understanding how the code works which is likely, will require another API change to pass the logging information through the datastore. I can see this becoming a very bloated system much more quickly than a global by comparison, hence the RFC nature of this PR to be a discussion point on how we want to handle a growing set of requirements in observability, from APM to logging to metrics.

Reply:
Yeah, I know it won't work which is why its an RFC :). As for using a `pool.Get(ctx)` pattern, I think that will be a requirement as go becomes more context based. Most libraries are changing to take in a context for this reason.

The main concern I have is two fold:

1. The redis library we use (written by you) is not just used by us. So making significant changes to remain 'globally pure' might have much farther reaching consequences.
2. While I agree with global purity, the very next change would be to enable logging in this area of code, which unless I am not understanding how the code works which is likely, will require another API change to pass the logging information through the datastore. I can see this becoming a very bloated system much more quickly than a global by comparison, hence the RFC nature of this PR to be a discussion point on how we want to handle a growing set of requirements in observability, from APM to logging to metrics.

Reply:
Actually, another idea would be to rework how async stuff is done so that its a first class router-like object. This would allow us to pass logging and APM info via a context much like how the HTTP router works currently.

Reply:
We don't really use `redigo` or my `redisc` package directly - both are abstracted to some degree in our `server/datastore/redis` package. That's what I meant when I say passing the tracing enabled flag when creating the Redis datastore, that doesn't require changes to either redigo or redisc.

I'm not a fan of over-using the context to pass around stuff, but since we're talking about "cross-cutting concerns" such as tracing/metrics, those are some of the things that make sense to pass via the context, as they are not really "hard dependencies" - the code works fine without those - but if provided they should be used to record whatever we want to record with them (that's also how OpenTelemetry works in Go IIRC). So in this case, it'd make sense to have the `main` setup the tracing and logging and store that in the context, and then when using `pool.Get(ctx)`, there is no further config to pass around, it just calls something like `tracer := somecontextpkg.TracerFromContext(ctx)` and uses it, and send a no-op tracer when it is disabled (IMO it is cleaner than having to check if it returned non-nil, but both work of course).

That being said, what would the logger be used for? Typically, it's cleaner (e.g. to avoid double-logging) when packages deep in the call stack return errors, and packages at the "higher" levels would handle logging (in that case, for example, the code that actually uses the redis connection would log if the command failed).

Reply:
> We don't really use redigo or my redisc package directly - both are abstracted to some degree in our server/datastore/redis package.

I tried adding this in there initially but doing it there seemed like it wouldn't be easy.

> I'm not a fan of over-using the context to pass around stuff,

I agree, but for this specific use case its a case where passing it via context makes sense.

> That being said, what would the logger be used for?

With observability (opentelemetry, apm, etc), log lines can be associated with spans. So we want logging to happen as close as possible to where the event ocured. This would be standard debug logging (strings for developers/humans to read). If there's already a way we handle that then we can handle it there but I didn't see any truly application wide logging framework that was available.

Reply:
> With observability (opentelemetry, apm, etc), log lines can be associated with spans.

Gotcha, so it sounds like it's more "contextual messages" to associate with spans, looks like it could be part of the same context-based approach. I know that opentelemetry at least allows associating names and attributes to spans, and named events (https://opentelemetry.io/docs/instrumentation/go/manual/#creating-spans), maybe our app logging could also be span-aware and associate its logs with the current span when there's one.

Reply:
> maybe our app logging could also be span-aware and associate its logs with the current span when there's one.

https://opentelemetry.io/docs/reference/specification/logs/overview/#log-correlation

https://www.elastic.co/guide/en/apm/agent/go/1.x/log-correlation.html

That's what I had in mind.